File: java_pure/src/main/java/util/net/retrofit/RetrofitApi.java
package util.net.retrofit;

        import kotlin.jvm.JvmSuppressWildcards;
        import okhttp3.MultipartBody;
        import okhttp3.RequestBody;
        import okhttp3.ResponseBody;
        import retrofit2.Call;
        import retrofit2.adapter.rxjava.Result;
        import retrofit2.http.*;

        import java.util.Map;

/**
 * Created by catfaceWYH --> tel|wechat|qq 130 128 92925
 *
 * @desc 封装风格与OkHttp一样(使用)
 */
public interface RetrofitApi {

    /* retrofit2 Missing either @GET URL or @Url parameter. --> @GET(".") or @GET("/") */
    @GET
    Call<String> get(@Url String suffix);

    /* 有参数 + @QueryMap */
    @GET
    Call<String> get(@Url String suffix, @QueryMap Map<String, String> map);


    @POST
    Call<String> post(@Url String suffix, @QueryMap Map<String, String> map);

    /* 有参数 + @FieldMap and @FormUrlEncoded */
    @FormUrlEncoded
    @POST
    Call<String> postField(@Url String suffix, @FieldMap Map<String, String> map);

    @Headers({"Content-Type: application/json", "Accept: application/json"})    // 需要添加头
    @POST
    Call<String> post(@Url String suffix, @Body RequestBody json);              // 传入的参数为RequestBody


    @Multipart
    @POST
    Call<String> upload(@Url String suffix, @Part MultipartBody.Part file);

    @JvmSuppressWildcards   // [IllegalArgumentException: Parameter type must not include a type variable or wildcard]
    @Multipart
    @POST
    Call<String> upload(@Url String suffix, @QueryMap Map<String, String> map, @PartMap Map<String, RequestBody> fileMap);

    @POST
    Call<String> upload(@Url String suffix, @Body RequestBody Body);


    @Streaming  // 避免下载大文件被Retrofit整个读进内存
    @GET
    Call<ResponseBody> download(@Url String suffix);


    /* 带进度的文件上传和下载 */
    @Multipart
    @POST
    Call<Result> uploadWithProgress(@Part RequestBody file);

}

File: java_pure/src/main/kotlin/net/retrofit/base/encapsulation/RetrofitApi.kt
package net.retrofit.base.encapsulation

import okhttp3.MultipartBody
import okhttp3.RequestBody
import okhttp3.ResponseBody
import retrofit2.Call
import retrofit2.http.*

/**
 * Created by catfaceWYH --> tel|wechat|qq 130 128 92925
 */
interface RetrofitApi {

    companion object {
        val HOST = "http://127.0.0.1:8080/"
    }


    /* 1. 仅请求服务器数据 */
    @GET
    fun get(@Url suffixUrl: String): Call<String>


    /* 2. 上传参数 */
    @POST
    fun post(@Url suffixUrl: String, @QueryMap map: Map<String, String>): Call<String>


    /* 3. ???上传json[暂时处于僵硬状态...] */
    /*@POST("postJson")
    function postJson(@Body user: User): Call<String>*/


    /* 4. 上传单个文件 */
    @Multipart
    @POST
    fun upload(@Url suffixUrl: String, @Part file: MultipartBody.Part): Call<String>


    /* 5. 上传多个文件 & 多个参数 */
    @JvmSuppressWildcards   // [IllegalArgumentException: Parameter type must not include a type variable or wildcard]
    @Multipart
    @POST
    fun upload(@Url suffixUrl: String, @QueryMap map: Map<String, String>, @PartMap fileMap: Map<String, RequestBody>): Call<String>


    /* 6. 下载文件 */
    @Streaming  // 避免下载大文件被Retrofit整个读进内存
    @GET
    fun download(@Url suffixUrl: String): Call<ResponseBody>
}